contracts/src/staking/README.md

```diff
@@ -0,0 +1,136 @@
+
+Once the Subnet has been recovered, the stake specified on the P-Chain must be locked in the Staking Contract before any further validator set changes may be made. This is done via the `provideInitialStake` method, and guarded by the `onlyWhenInitialStakeProvided` modifier.
+
+
+### Registering a Validator
+
+Validator registration is initiated with a call to `initializeValidatorRegistration` on the concrete Staking Manager contract. The Staking Manager contract constructs a [`RegisterSubnetValidatorMessage`](#registersubnetvalidatormessage) Warp message to be sent to the P-Chain. Each validator registration request includes all of the information needed to identify the validator and its stake weight, as well as an `expiry` timestamp before which the `RegisterSubnetValidatorMessage` must be delivered to the P-Chain.
+
+The `RegisterSubnetValidatorMessage` is delivered to the P-Chain as the payload of a `RegisterSubnetValidatorTx`. Please see the transaction [specification](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#step-2-issue-a-registersubnetvalidatortx-on-the-p-chain) for validity requirements. The successful `RegisterSubnetValidatorTx` results in a [`SubnetValidatorRegistrationMessage`](#subnetvalidatorregistrationmessage) Warp message indicating that the specified validation was successfully registered on the P-Chain.
+
+The `SubnetValidatorRegistrationMessage` is delivered to the Staking Manager contract via a call to `completeValidatorRegistration`. Staking rewards begin accruing at this time.
+
+#### Handling a Missed Expiry
+
+In the case of a missed expiry, the `RegisterSubnetValidatorTx` will result in a `SubnetValidatorRegistrationMessage` Warp message with the `valid` field set to `0`. This serves as proof that the corresponding validation has not started and may never start. The `SubnetValidatorRegistrationMessage` can be provided to the Staking Manager contract by calling `completeEndValidation` with `setWeightMessageType=false`.
+
+### Removing a Validator
+
+As an optional first step, an uptime proof may be requested from the P-Chain, which will be used the calculate the staking rewards on the Subnet. The proof is embedded in a [`ValidationUptimeMessage`](#validationuptimemessage) Warp message.
+
+Validators exit is initiated with a call to `initializeEndValidation` on the Staking Manager contract. A `ValidationUptimeMessage` Warp message may optionally be provided, otherwise the validator's uptime will be set to`0`. Once `initializeEndValidation` staking rewards cease accruing. The Staking Manager contract constructs a [`SetSubnetValidatorWeightMessage`](#setsubnetvalidatorweightmessage) Warp message, setting the weight to `0`.
```

Comment:
Just thinking it would probably make sense to provide two external functions - one that expects rewards to be disbursed, and one that exits without rewards. This would avoid someone accidentally not including their uptime proof, which could be a costly mistake.

Reply:
Adding to this, my understanding (which may be incorrect) is that Warp messages can be delivered out of order. If this is correct, then is there a chance that the SetSubnetValidatorWeightMessage message be delivered before the ValidationUptimeMessage message?

Reply:
An alternative would be to update the redeemable amount internally on validation completion, and have disbursement be a separate step, similar to `redeemRelayerRewards` in `TeleporterMessenger`.

Reply:
> is there a chance that the SetSubnetValidatorWeightMessage message be delivered before the ValidationUptimeMessage message?

This is only possible if `completeEndValidation` is called before `initializeEndValidation`, in which case the default uptime value of 0 will be used.